sdk/storage/azure-storage-file-share/src/main/java/com/azure/storage/file/share/ShareFileAsyncClient.java

```diff
@@ -1310,18 +1322,62 @@ private void channelCleanUp(AsynchronousFileChannel channel) {
      * @throws ShareStorageException If you attempt to upload a range that is larger than 4 MB, the service returns
      * status code 413 (Request Entity Too Large)
      */
-    @ServiceMethod(returns = ReturnType.SINGLE)
+    @ServiceMethod(returns = ReturnType.SINGLE) //TODO communicate no longer using length
```

Comment:
It may be better to leave this method as-is given the runtime behavior change being made to it. Similar to how BlockBlobAsyncClient has an API which only performs the single shot upload and then has the length agnostic uploads.

Reply:
The rest of our SDKs use "upload" to name methods that do the work for you and we use specialized names for more specific methods. What are you naming suggestions for such a new method that wouldn't do a 180 and confuse users?

Reply:
@alzimmermsft What do you think of leaving the behavior of these methods as is and deprecating them in favor of some new methods called uploadRange (this would match well with existing clearRange) and we can create new APIs for buffered upload as follows

async
upload(Flux, PTO)
upload(Flux, PTO, boolean)
uploadWithResponse(UploadOptions)

sync
upload(IS, long)
upload(IS, long, boolean)
uploadWithResponse(UploadOptions, Duration, Context)

Reply:
That is roughly what I was thinking. Leaving the functionality of the current APIs as-is and introducing new APIs which explicitly support any length input and handles chunking internally.

Reply:
Only weirdness with this is that we already  have a method in sync called upload(IS, long)

Reply:
In .net we had a similar situation with blobs where we deprecated a whole method to introduce two newly named ones, when one of them was exactly the same as the deprecated one. Download became DownloadStreaming and DownloadContent. DownloadStreaming was just a new name for Download, but now there was obvious connection and distinction between method names.

We could do that here. Completely scrap Upload and have two completely new names. I'm struggling to think of what those names would be, though. They both take ranges, can both upload to a specific offset. Theonly difference is that one needs a length parameter because we don't buffer while the other has no range size cap.

Reply:
`UploadRange(data, length, ...)` and `UploadBufferedRange(data, ...)`?

Reply:
@alzimmermsft we've setup a new strategy for how to handle this, which is detailed in more recent comments.

Reply:
The new pattern looks great to me, I'm happy to see (though its less desirable) that the current, soon to be previous, uploading APIs are maintain runtime behavior of only allowing a 4MB chunk at max and new APIs were added for the unbounded upload size.